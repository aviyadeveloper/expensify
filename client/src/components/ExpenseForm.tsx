import React from 'react';
import moment, { Moment } from 'moment';
import { SingleDatePicker } from 'react-dates';
import { Expense } from '../types/expensesTypes';
import { isNull } from 'util';

type ExpenseFormProps = {
  onSubmit?: (expense: Expense) => void;
  name?: string;
  description?: string;
  amount?: number;
  createdAt?: string;
  calendarFocused?: boolean | null;
  error?: string;
  editMode?: boolean;
};

interface ExpenseFormState {
  name: string;
  description: string;
  amount: string;
  createdAt: string;
  calendarFocused: boolean | null;
  error: string;
  editMode?: boolean;
}

export class ExpenseForm extends React.Component<
  ExpenseFormProps,
  ExpenseFormState
> {
  state: ExpenseFormState = {
    name: this.props.name || '',
    description: this.props.description || '',
    amount: (this.props.amount && (this.props.amount / 100).toFixed(2)) || '',
    createdAt:
      (this.props.createdAt && moment(this.props.createdAt).format()) ||
      moment().format(),
    calendarFocused: false,
    error: '',
    editMode: this.props.editMode
  };

  onNameChange = (e: React.FormEvent<HTMLInputElement>) => {
    const name = e.currentTarget && e.currentTarget.value;
    this.setState({ name });
  };

  onAmountChange = (e: React.FormEvent<HTMLInputElement>) => {
    let amount = e.currentTarget && e.currentTarget.value;

    // Allow only cent based currency fit number format.
    (amount.match(/^[1-9]\d*(\.\d{0,2})?$/) || !amount) &&
      this.setState({ amount });
  };

  onDateChange = (newDate: Moment | null): void => {
    !isNull(newDate) && this.setState({ createdAt: newDate.format() });
  };

  onCalendarFocused = ({ focused }: { focused: boolean | null }) => {
    this.setState(() => ({ calendarFocused: focused }));
  };

  onDescriptionChange = (e: React.FormEvent<HTMLTextAreaElement>) => {
    let description = e.currentTarget && e.currentTarget.value;
    this.setState({ description });
  };

  onSubmit = (event: React.FormEvent<HTMLFormElement>) => {
    event.preventDefault();

    if (!this.state.amount || !this.state.name) {
      this.setState(() => ({ error: 'Name and amount are required' }));
    } else {
      this.setState(() => ({ error: '' }));
      {
        this.props.onSubmit &&
          this.props.onSubmit({
            name: this.state.name,
            amount: parseFloat(this.state.amount) * 100,
            description: this.state.description,
            createdAt: this.state.createdAt
          });
      }
    }
  };

  render() {
    return (
      <div className="content-container">
        <form className="input-group-vertical" onSubmit={this.onSubmit}>
          <div className="input-group__item">
            <input
              id="expense-form-name"
              className="input-group__input"
              type="text"
              placeholder="name"
              autoFocus
              value={this.state.name}
              onChange={this.onNameChange}
            />
          </div>
          <div className="input-group__item">
            <input
              id="expense-form-amount"
              className="input-group__input"
              type="text"
              placeholder="amount"
              value={this.state.amount}
              onChange={this.onAmountChange}
            />
          </div>
          <div className="input-group__item">
            <SingleDatePicker
              id="expense-form-date"
              date={moment()}
              onDateChange={this.onDateChange}
              focused={!!this.state.calendarFocused}
              onFocusChange={this.onCalendarFocused}
              numberOfMonths={1}
              isOutsideRange={() => false}
            />
          </div>
          <div className="input-group__item">
            <textarea
              id="expense-form-description"
              className="input-group__input-long"
              value={this.state.description}
              onChange={this.onDescriptionChange}
              placeholder="description"
            />
          </div>
          {this.state.error && <p>{this.state.error}</p>}
          <button className="buttons__regular">
            {this.props.editMode ? 'Edit' : 'Add'}
          </button>
        </form>
      </div>
    );
  }
}
