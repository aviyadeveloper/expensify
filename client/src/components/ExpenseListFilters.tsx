import React, { FormEvent } from 'react';
import { connect } from 'react-redux';
import {
  setNameFilter,
  sortByDate,
  sortByAmount,
  setStartDate,
  setEndDate
} from '../actions/filtersActions';
import { FiltersSortBy } from '../types/filtersTypes';
import { DateRangePicker } from 'react-dates';
import { Moment } from 'moment';
import { Dispatch } from 'redux';

export class ExpenseListFilters extends React.Component<any, any> {
  state = {
    focusedInput: null
  };

  onDatesChange = ({
    startDate,
    endDate
  }: {
    startDate: Moment | null;
    endDate: Moment | null;
  }) => {
    this.props.setStartDate(startDate);
    this.props.setEndDate(endDate);
  };

  onFocusChange = (focusedInput: 'startDate' | 'endDate' | null) => {
    this.setState({ focusedInput });
  };

  onNameChange = (event: FormEvent<HTMLInputElement>) => {
    this.props.setNameFilter(event.currentTarget.value || '');
  };

  onFilterTypeChange = (event: FormEvent<HTMLSelectElement>) => {
    const value: number = parseInt(event.currentTarget.value);
    switch (value) {
      case FiltersSortBy.Date:
        return this.props.sortByDate();
      case FiltersSortBy.Amount:
        return this.props.sortByAmount();
      default:
        return this.props.sortByDate();
    }
  };

  render() {
    return (
      <div className="content-container">
        <div className="input-group">
          <div className="input-group__item">
            <input
              id="expense-filter-name"
              className="input-group__input"
              placeholder="Filter by name"
              value={this.props.filters.name}
              onChange={this.onNameChange}
            />
          </div>
          <div className="input-group__item">
            <select
              id="expense-filter-sortBy"
              className="input-group__input"
              value={this.props.filters.sortBy}
              onChange={this.onFilterTypeChange}
            >
              <option value={FiltersSortBy.Date}>Date</option>
              <option value={FiltersSortBy.Amount}>Amount</option>
            </select>
          </div>
          <div className="input-group__item">
            <DateRangePicker
              startDate={this.props.filters.startDate}
              startDateId={'sortByDate-filter-startDate'}
              endDate={this.props.filters.endDate}
              endDateId={'sortByDate-filter-endDate'}
              onDatesChange={this.onDatesChange}
              focusedInput={this.state.focusedInput}
              onFocusChange={this.onFocusChange}
              isOutsideRange={() => false}
              numberOfMonths={1}
              showClearDates={true}
            />
          </div>
        </div>
      </div>
    );
  }
}

const mapStateToProps = (state: any) => ({
  filters: state.filters
});

const mapDispatchToProps = (dispatch: Dispatch) => ({
  setNameFilter: (name: string) => {
    dispatch(setNameFilter(name));
  },
  sortByAmount: () => {
    dispatch(sortByAmount());
  },
  sortByDate: () => {
    dispatch(sortByDate());
  },
  setStartDate: (startDate: Moment | null) => {
    dispatch(setStartDate(startDate));
  },
  setEndDate: (endDate: Moment | null) => {
    dispatch(setEndDate(endDate));
  }
});
export default connect(
  mapStateToProps,
  mapDispatchToProps
)(ExpenseListFilters);
