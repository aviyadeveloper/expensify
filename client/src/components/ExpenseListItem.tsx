import React from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import { Expense } from '../types/expensesTypes';
import moment from 'moment';
import numeral from 'numeral';

export const ExpenseListItem = (props: Expense) => (
  <div>
    <Link to={`/edit/${props.id}`}>
      <h5>{props.name}</h5>
    </Link>
    <p>{props.description}</p>
    <p>
      <span>
        {numeral(props.amount && props.amount / 100).format('$0,0.00')}
      </span>{' '}
      / <span>{moment(props.createdAt).format('MMMM Do, YYYY')}</span>
    </p>
  </div>
);

export default connect()(ExpenseListItem);
