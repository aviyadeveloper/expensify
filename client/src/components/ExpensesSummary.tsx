import React from 'react';
import { connect } from 'react-redux';
import expensesTotalSelector from '../selectors/expensesTotalSelector';
import numeral from 'numeral';
import { Expense } from '../types/expensesTypes';
import { Link } from 'react-router-dom';

type ExpensesSummaryProps = {
  expenses: Expense[];
};

export const ExpensesSummary = (props: ExpensesSummaryProps) => (
  <div className="page-header">
    <div className="content-container">
      <h4 className="page-header__title">
        Viewing <span>{props.expenses.length}</span> expense
        {props.expenses.length !== 1 && 's'} totalling in{' '}
        <span>
          {numeral(expensesTotalSelector(props.expenses) / 100).format(
            '$0,0.00'
          )}
        </span>
      </h4>
      <Link to="add" className="buttons__regular">
        Add Expense
      </Link>
    </div>
  </div>
);

export default connect()(ExpensesSummary);
