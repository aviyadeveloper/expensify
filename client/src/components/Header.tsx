import React from 'react';
import { connect } from 'react-redux';
import { withRouter, RouteComponentProps } from 'react-router';
import { NavLink, Link } from 'react-router-dom';
import { runLogout } from '../actions/auth';

interface IHeaderProps extends RouteComponentProps {
  runLogout?: () => void;
}

export const Header: React.SFC<IHeaderProps> = props => (
  <header className="navbar">
    <Link className="navbar__title-link-container" to="/dashboard">
      <h1 className="navbar__app-title">Expensify</h1>
    </Link>
    <div className="navbar__links-container">
      <NavLink
        activeClassName="active-link"
        className="navbar__links-container__link"
        to="/help"
      >
        Help
      </NavLink>
      <button
        id="logout-button"
        className="navbar__links-container__link buttons__clear"
        onClick={props.runLogout}
      >
        Logout
      </button>
    </div>
  </header>
);

const mapDispatchToProps = (dispatch: Function) => ({
  runLogout: () => {
    dispatch(runLogout());
  }
});

export default withRouter(
  connect(
    undefined,
    mapDispatchToProps
  )(Header)
);
