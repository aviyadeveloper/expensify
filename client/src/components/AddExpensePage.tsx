import React, { Dispatch } from 'react';
import { connect } from 'react-redux';
import { ExpenseForm } from './ExpenseForm';
import { Expense } from '../types/expensesTypes';
import { runAddExpense } from '../actions/expensesActions';
import { History } from 'history';

type AddExpensePageProps = {
  runAddExpense: (expense: Expense) => void;
  history: History;
};

export class AddExpensePage extends React.Component<AddExpensePageProps> {
  runAddExpense = (expense: Expense): void => {
    this.props.runAddExpense(expense);
    this.props.history.push('/dashboard');
  };
  render() {
    return (
      <div>
        <div className="page-header">
          <div className="content-container">
            <h2 className="page-header__title">Add New Expense</h2>
          </div>
        </div>

        <ExpenseForm onSubmit={this.runAddExpense} />
      </div>
    );
  }
}

const mapDispatchToProps = (dispatch: Dispatch<Function>) => ({
  runAddExpense: (expense: Expense) => dispatch(runAddExpense(expense))
});

export default connect<any>(
  undefined,
  mapDispatchToProps
)(AddExpensePage);
