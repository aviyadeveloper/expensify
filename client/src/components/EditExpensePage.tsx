import React, { Dispatch } from 'react';
import { ExpenseForm } from './ExpenseForm';
import { Expense } from '../types/expensesTypes';
import { connect } from 'react-redux';
import { runEditExpense, runRemoveExpense } from '../actions/expensesActions';
import { History } from 'history';
import { match } from 'react-router';
import { storeStateType } from '../types/storeTypes';

export type EditExpensePageProps = {
  runEditExpense: (id: string, expense: Expense) => void;
  runRemoveExpense: (id: string) => void;
  match: match<{ id: string | undefined }>;
  expense: Expense | undefined;
  history: History;
};

export class EditExpensePage extends React.Component<EditExpensePageProps> {
  onEditExpense = (expense: Expense) => {
    if (this.props.expense && typeof this.props.expense.id === 'string') {
      this.props.runEditExpense(this.props.expense.id, expense);
      this.props.history.push('/dashboard');
    }
  };

  onRemoveExpense = () => {
    if (this.props.expense && typeof this.props.expense.id === 'string') {
      this.props.runRemoveExpense(this.props.expense.id);
      this.props.history.push('/dashboard');
    }
  };

  render() {
    return (
      <div>
        <div className="page-header">
          <div className="content-container">
            <div className="page-header__top">
              <h2 className="page-header__title">Editing Expense</h2>
              <button
                id="remove-expense-button"
                className="buttons__regular"
                onClick={this.onRemoveExpense}
              >
                Delete Expense
              </button>
            </div>
          </div>
        </div>
        <ExpenseForm
          onSubmit={this.onEditExpense}
          editMode={true}
          {...this.props.expense}
        />
      </div>
    );
  }
}

const mapStateToProps = (
  state: storeStateType,
  props: EditExpensePageProps
) => ({
  expense: state.expenses.find(
    (expense: Expense): boolean => expense.id === props.match.params.id
  )
});

const mapDispatchToProps = (dispatch: Dispatch<any>) => ({
  runEditExpense: (id: string, expense: Expense) => {
    dispatch(runEditExpense(id, expense));
  },
  runRemoveExpense: (id: string) => {
    dispatch(runRemoveExpense(id));
  }
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(EditExpensePage);
