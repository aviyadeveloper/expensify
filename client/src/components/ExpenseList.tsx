import React from "react";
import { connect } from "react-redux";
import { Expense } from "../types/expensesTypes";
import { FiltersReducerState } from "../types/filtersTypes";
import ExpenseListItem from "./ExpenseListItem";
import ExpenseListFilters from "./ExpenseListFilters";
import filterExpenses from "../selectors/expensesSelector";

type ExpenseListStateProps = {
  expenses: Expense[];
  filters: FiltersReducerState;
};

export const ExpenseList = (props: ExpenseListStateProps) => (
  <div>
    <ExpenseListFilters />
    <div>
      {props.expenses.length < 1 ? (
        <p>No expenses found.</p>
      ) : (
        props.expenses.map(e => <ExpenseListItem key={e.id} {...e} />)
      )}
    </div>
  </div>
);

const mapStateToProps = (state: ExpenseListStateProps) => ({
  expenses: filterExpenses(state.expenses, state.filters)
});

export default connect(mapStateToProps)(ExpenseList);
