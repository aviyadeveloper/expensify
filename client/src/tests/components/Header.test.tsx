import React from 'react';
import { shallow } from 'enzyme';
import { Header } from '../../components/Header';
import {
  createBrowserHistory,
  createLocation,
  createPath,
  History,
  Location
} from 'history';
import { match } from 'react-router-dom';

interface HeaderTestProps {
  runLogout: () => void;
  history: History;
  location: Location;
  match: match;
}
let HeaderTestProps: HeaderTestProps;

beforeEach(() => {
  HeaderTestProps = {
    runLogout: jest.fn(),
    history: createBrowserHistory(),
    location: createLocation(createPath({})),
    match: { params: {}, isExact: false, path: '', url: '' }
  };
});

test('should render Header component', () => {
  const wrapper = shallow(<Header {...HeaderTestProps} />);
  expect(wrapper).toMatchSnapshot();
});

test('should invoke runLogout when logout button clicked', () => {
  const wrapper = shallow(<Header {...HeaderTestProps} />);
  wrapper.find('button#logout-button').simulate('click');
  expect(HeaderTestProps.runLogout).toHaveBeenCalled();
});
