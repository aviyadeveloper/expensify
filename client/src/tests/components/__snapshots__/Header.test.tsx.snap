// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Header component 1`] = `
<header
  className="navbar"
>
  <div
    className="content-container"
  >
    <Link
      className="navbar__title-link-container"
      replace={false}
      to="/dashboard"
    >
      <h1
        className="navbar__app-title"
      >
        Expensify
      </h1>
    </Link>
    <div
      className="navbar__links-container"
    >
      <NavLink
        activeClassName="active-link"
        aria-current="page"
        className="navbar__links-container__link"
        to="/help"
      >
        Help
      </NavLink>
      <button
        className="navbar__links-container__link buttons__clear"
        id="logout-button"
        onClick={[MockFunction]}
      >
        Logout
      </button>
    </div>
  </div>
</header>
`;
